File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        var node1 = l1;
        var node2 = l2;

        ListNode sumNode = null;
        ListNode sumHead = null;
        int carry = 0;

        while (node1 != null || node2 != null || carry > 0) {
            
            var sum = carry;

            if (node1 != null) {
                sum += node1.val;
            }

            if (node2 != null) {
                sum += node2.val;
            }

            carry = sum / 10;
            sum %= 10;

            final var newNode = new ListNode(sum);
            if (sumNode != null) {
                sumNode.next = newNode;
            }

            sumNode = newNode;

            if (sumHead == null) {
                sumHead = sumNode;
            }

            node1 = node1 == null ? null : node1.next;
            node2 = node2 == null ? null : node2.next;
        }

        return sumHead;
    }
}

File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if not l1 or not l2:
            return l1 or l2
        
        head, prev_node = None, None
        carry = 0
        
        pointer1, pointer2 = l1, l2
        
        while carry > 0 or pointer1 or pointer2:
            
            sum = carry
            
            if pointer1:
                sum += pointer1.val
                pointer1 = pointer1.next
                
            if pointer2:
                sum += pointer2.val
                pointer2 = pointer2.next
            
            carry = sum // 10
            
            new_node = ListNode(val=sum % 10)
            
            if prev_node:
                prev_node.next = new_node
            else:
                head = new_node
                
            prev_node = new_node
        
        return head